gossipsub.go

```diff
@@ -441,16 +553,21 @@ func (gs *GossipSubRouter) heartbeat() {
 	tograft := make(map[peer.ID][]string)
 	toprune := make(map[peer.ID][]string)
 
+	// clean up expired backoffs
+	gs.clearBackoff()
```

Comment:
This is probably fine, but we should monitor this.

Reply:
That is, walking through each of these every heartbeat _should_ be fine, but could get expensive in some edge cases.

Reply:
We could do it every few heartbeats instead of every heartbeat.

Reply:
Every heartbeat is probably fine, I'm just calling it out so we keep it in mind.

Reply:
A way more efficient manner would be to use a ring buffer of peer slices, with `nslots = BackoffPeriod/HeartbeatInterval`. Then you track the tick count (monotonically increasing by 1), every heartbeat increments it by 1. And when clearing the backoff, you simply do `currentTick mod nslots`, and simply clear that slot. That's pretty efficient. The data for backoff would turn into `map[string][][]peer.ID`.